### PR TITLE
Sigmoid function

### PR DESCRIPTION
**Pull request by sgasioro**
Add in sigmoid function as discussed [here](https://indico.slac.stanford.edu/event/7170/contributions/3667/attachments/1554/4064/Differentiability%20of%20Sim%20Pieces.pdf). Includes a switch between smooth and not smoothed versions of simulation, with ~ reasonable defaults set.

Some small additional edits: make types match in `all_sim`, pass vdrift as an argument to checkpointed function to allow gradient flow to work, add differentiable versions of arange and linspace to a utils.py (but these don't seem to matter (?) so not used anywhere).

### CODE REVIEW

File: larndsim/detsim_ep.py [hunk 1]
```diff
@@ -132,6 +132,10 @@ def z_interval(self, start_point, end_point, x_p, y_p, tolerance, eps=1e-12):
     def erf_hack(self, input):
         return ep.astensor(torch.erf(input.raw))
 
+    def sigmoid(self, x):
+        # Using torch here, otherwise we get overflow -- might be able to get around with a trick
+        return ep.astensor(torch.sigmoid(x.raw))
```
Comment (ynashed) [inline larndsim/detsim_ep.py:137]: Oh no! Another place to remember when de-pytorchizing. Oh well.

File: larndsim/detsim_ep.py [hunk 2]
```diff
@@ -291,11 +300,10 @@ def calc_total_current(self, x_start, y_start, z_start,
 
         z_step = (z_end_int - z_start_int) / (z_steps - 1)
 
-
         iz = ep.arange(z_steps, 0, z_steps.max().item())
         z =  z_start_int[:, :, ep.newaxis] + iz[ep.newaxis, ep.newaxis, :] * z_step[..., ep.newaxis]
 
-        t0 = (ep.abs(z - borders[:, 2, 0, ep.newaxis, ep.newaxis]) - 0.5) / self.vdrift
+        t0 = (ep.abs(z - borders[:, 2, 0, ep.newaxis, ep.newaxis]) - 0.5) / vdrift
```
Comment (ynashed) [inline larndsim/detsim_ep.py:306]: What's the explanation for this change?

File: larndsim/detsim_ep.py (hunk 1 above)
Comment (sgasioro) [reply larndsim/detsim_ep.py:137]: Yeah, you can avoid this by splitting into x > 0 and x < 0 cases and just writing the function in different ways ([here's a Medium article](https://shaktiwadekar.medium.com/how-to-avoid-numerical-overflow-in-sigmoid-function-numerically-stable-sigmoid-function-5298b14720f6) with terrible screenshots lol), but need to think about how to do this in our case

File: larndsim/utils.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    start = ep.astensor(start)
+    end = ep.astensor(end)
+    num = int(num)
+    
+    delta = (end - start) / num
+    if endpoint:
+        delta *= num / (num - 1)
+    return start + ep.arange(end, 0, num)*delta
+
+def diff_arange(start, end):
```
Comment (ynashed) [inline larndsim/utils.py:13]: Nice 👍

**Review by ynashed — APPROVED** [review]

File: larndsim/detsim_ep.py (hunk 2 above)
Comment (sgasioro) [reply larndsim/detsim_ep.py:306]: Torch checkpointing only keeps track of gradients for explicit inputs and outputs of the checkpointed function, so if you don't pass vdrift as an argument and vdrift has `requires_grad=True` it complains. Not the cleanest, but so it goes
